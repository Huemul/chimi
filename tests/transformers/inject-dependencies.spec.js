const { stripIndent } = require('common-tags')

const injectDependencies = require('../../lib/transformers/inject-dependencies')

describe('injectDependencies', () => {
  const code = stripIndent`
    const answer = 42

    console.log(answer)
  `

  const position = {
    start: { line: 0, column: 0 },
  }

  it('should add source maps but no dependencies', () => {
    const result = injectDependencies(position, [], {})('README.md', code)

    expect(result.code).toMatchSnapshot()
  })

  it('should not change snippet indentation', () => {
    const withIndentation = stripIndent`
      const foo = {
        bar: "bar"
      };

      Promise.resolve(foo)
        .then(({ bar }) => bar)
        .then(bar => {
          console.log(bar);
        });
    `
    let result = injectDependencies(position, [], {})(
      'README.md',
      withIndentation
    )

    expect(result.code).toMatchSnapshot()

    const dependencies = [
      {
        name: '_',
        module: 'lodash',
      },
      {
        name: 'lib',
        module: './lib',
      },
    ]
    result = injectDependencies(position, dependencies, {})(
      'README.md',
      withIndentation
    )

    expect(result.code).toMatchSnapshot()
  })

  it('should prepend requires with variable declarations to the snippet code', () => {
    let dependencies = [
      {
        name: '_',
        module: 'lodash',
      },
    ]
    let result = injectDependencies(position, dependencies, {})(
      'README.md',
      code
    )

    expect(result.code).toMatchSnapshot()

    dependencies = [
      {
        name: '_',
        module: 'lodash',
      },
      'trae',
    ]
    result = injectDependencies(position, dependencies, {})('README.md', code)

    expect(result.code).toMatchSnapshot()
  })

  it('should work for local dependencies', () => {
    const dependencies = [
      {
        name: 'lib',
        module: './lib',
      },
      {
        name: 'bar',
        module: '../foo/bar',
      },
    ]
    const result = injectDependencies(position, dependencies, {})(
      'README.md',
      code
    )

    expect(result.code).toMatchSnapshot()
  })

  it('should add requires with no declaration when the name is missing', () => {
    const dependencies = [
      {
        module: './for-the-side-effects',
      },
      {
        module: 'es6-promise',
      },
    ]
    const result = injectDependencies(position, dependencies, {})(
      'README.md',
      code
    )

    expect(result.code).toMatchSnapshot()
  })

  it('should add globals', () => {
    const globals = {
      theAnswer: 42,
    }
    const result = injectDependencies(position, [], globals)('README.md', code)

    expect(result.code).toMatchSnapshot()
  })

  it('should add globals and dependencies', () => {
    const dependencies = [
      {
        name: '_',
        module: 'lodash',
      },
    ]
    const globals = {
      theAnswer: 42,
    }
    const result = injectDependencies(position, dependencies, globals)(
      'README.md',
      code
    )

    expect(result.code).toMatchSnapshot()
  })

  it('should interpolate functions, arrays, objets and strings properly', () => {
    const globals = {
      add: '(a, b) => a + b',
      message: "'hello world'",
      foo: "{ foo: 'bar' }",
      fruits: "['orange', 'apple']",
    }
    const result = injectDependencies(position, [], globals)('README.md', code)

    expect(result.code).toMatchSnapshot()
  })
})
