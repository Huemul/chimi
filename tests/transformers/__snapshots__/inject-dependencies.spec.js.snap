// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectDependencies should add globals 1`] = `
"require('source-map-support').install()
// snippet dependencies

let theAnswer = 42;

const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should add globals and dependencies 1`] = `
"require('source-map-support').install()
// snippet dependencies
const _ = require('lodash');
let theAnswer = 42;

const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should add requires with no declaration when the name is missing 1`] = `
"require('source-map-support').install()
// snippet dependencies
require('./for-the-side-effects');
require('es6-promise');


const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should add source maps but no dependencies 1`] = `
"require('source-map-support').install()
// snippet dependencies



const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should interpolate functions, arrays, objets and strings properly 1`] = `
"require('source-map-support').install()
// snippet dependencies

let add = (a, b) => a + b;
let message = 'hello world';
let foo = { foo: 'bar' };
let fruits = ['orange', 'apple'];

const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should not change snippet indentation 1`] = `
"require('source-map-support').install()
// snippet dependencies



const foo = {
  bar: \\"bar\\"
};

Promise.resolve(foo)
  .then(({ bar }) => bar)
  .then(bar => {
    console.log(bar);
  });"
`;

exports[`injectDependencies should not change snippet indentation 2`] = `
"require('source-map-support').install()
// snippet dependencies
const _ = require('lodash');
const lib = require('./lib');


const foo = {
  bar: \\"bar\\"
};

Promise.resolve(foo)
  .then(({ bar }) => bar)
  .then(bar => {
    console.log(bar);
  });"
`;

exports[`injectDependencies should prepend requires with variable declarations to the snippet code 1`] = `
"require('source-map-support').install()
// snippet dependencies
const _ = require('lodash');


const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should prepend requires with variable declarations to the snippet code 2`] = `
"require('source-map-support').install()
// snippet dependencies
const _ = require('lodash');
const trae = require('trae');


const answer = 42

console.log(answer)"
`;

exports[`injectDependencies should work for local dependencies 1`] = `
"require('source-map-support').install()
// snippet dependencies
const lib = require('./lib');
const bar = require('../foo/bar');


const answer = 42

console.log(answer)"
`;
